src/app/ReadClient.cpp

```diff
@@ -540,10 +540,27 @@ CHIP_ERROR ReadClient::ProcessEventReportIBs(TLV::TLVReader & aEventReportIBsRea
 
 CHIP_ERROR ReadClient::RefreshLivenessCheckTimer()
 {
+    CHIP_ERROR err = CHIP_NO_ERROR;
     CancelLivenessCheckTimer();
-    ChipLogProgress(DataManagement, "Refresh LivenessCheckTime with %d seconds", mMaxIntervalCeilingSeconds);
-    CHIP_ERROR err = InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager()->SystemLayer()->StartTimer(
-        System::Clock::Seconds16(mMaxIntervalCeilingSeconds), OnLivenessTimeoutCallback, this);
+    VerifyOrReturnError(mpExchangeCtx != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    System::Clock::Timeout timeout = System::Clock::Seconds16(mMaxIntervalCeilingSeconds);
+    if (mpExchangeCtx->IsUDPTransport())
+    {
+        auto margin = System::Clock::Milliseconds32(
+            (CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS + 1) *
+            (mpExchangeCtx->GetIdleRetransmitTimeoutTick() << CHIP_CONFIG_RMP_TIMER_DEFAULT_PERIOD_SHIFT));
```

Comment:
How exactly was this tested manually? Can you please describe the specific tests you did?

Reply:
I use im mock initiator and im mock responder, and after subscription is estabalished, I tear down responder, then check the print out in initiator and see if the actual timeout call can be triggered after max(5s) + margin(19seconds)

Reply:
[1637379765276] [44720:9993776] CHIP: [EM] Handling via exchange: 21819r, Delegate: 0x0x107376830
[1637379765276] [44720:9993776] CHIP: [DMG] ReportDataMessage =
[1637379765276] [44720:9993776] CHIP: [DMG] {
[1637379765276] [44720:9993776] CHIP: [DMG] 	SubscriptionId = 0x72bde65c628c1f48,
[1637379765276] [44720:9993776] CHIP: [DMG] }
[1637379765276] [44720:9993776] CHIP: [DMG] **Refresh LivenessCheckTime with 24968 milliseconds**
[1637379765276] [44720:9993776] CHIP: [EM] Piggybacking Ack for MessageCounter:18 on exchange: 21819r
[1637379765276] [44720:9993776] CHIP: [IN] Prepared encrypted message 0x107375658 to 0x0000000000BC5C01 of type 0x1 and protocolId (0, 1) on exchange 21819r with MessageCounter:25.
[1637379765276] [44720:9993776] CHIP: [IN] Sending encrypted msg 0x107375658 with MessageCounter:25 to 0x0000000000BC5C01 at monotonic time: 977658146 msec
[1637379765277] [44720:9993775] CHIP: [EM] Received message of type 0x10 with protocolId (0, 0) and MessageCounter:19 on exchange 21819r
[1637379765277] [44720:9993775] CHIP: [EM] Rxd Ack; Removing MessageCounter:25 from Retrans Table on exchange 21819r
[1637379765277] [44720:9993775] CHIP: [EM] Removed CHIP MessageCounter:25 from RetransTable on exchange 21819r
[1637379790244] [44720:9993776] CHIP: [DMG] Subscription Liveness timeout with peer node 0xbc5c01, shutting down 
ReadError with err ../../src/app/ReadClient.cpp:580: CHIP Error 0x00000032: Timeout[1637379790245] [44720:9993776] CHIP: [DMG] Client[0] moving to [UNINIT]